### PR TITLE
UCT/DC: initial implementation

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please review

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2000/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1376/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1378/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2002/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2003/ for details.

**Comment by alex-mikheev** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1379/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2004/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1380/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1384/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2006/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1385/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2007/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1386/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2008/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1388/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2012/ for details.

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2017/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1393/ for details (Mellanox internal link).

File: src/uct/ib/dc/base/dc_iface.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+typedef uct_ib_qpnum_t uct_dc_iface_addr_t;
+typedef struct uct_dc_ep uct_dc_ep_t;
+
+/* TODO: derive from rc config */
+typedef struct uct_dc_iface_config {
+    /* work around to do multiple inheritance:
+     * dc_verbs needs both dc_iface_config and verbs_common_iface config 
+     */
+    uct_rc_verbs_iface_config_t   super;
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.h]: pls indent struct fields on column
https://github.com/openucx/ucx/blob/master/doc/CodeStyle#L9

File: src/uct/ib/dc/base/dc_iface.c [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    ucs_free(self->tx.dcis_stack);
+    ucs_arbiter_cleanup(&self->tx.dci_arbiter);
+}
+
+
+UCS_CLASS_DEFINE(uct_dc_iface_t, uct_rc_iface_t);
+
+ucs_config_field_t uct_dc_iface_config_table[] = {
+    {"DC_", "", NULL,
+        ucs_offsetof(uct_dc_iface_config_t, super), UCS_CONFIG_TYPE_TABLE(uct_rc_verbs_iface_config_table)},
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: missing definitions of the configuration fields

File: src/uct/ib/dc/base/dc_iface.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+typedef struct uct_dc_dci {
+    uct_rc_txqp_t   txqp;
+    uct_dc_ep_t     *ep;        /* points to an endpoint that currently own
+                                   the dci */
+} uct_dc_dci_t;
+
+typedef struct uct_dc_iface { 
+    uct_rc_iface_t super;
+    struct {
+        uct_dc_dci_t  *dcis;
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.h]: pls add comments to the fields and how they are used

File: src/uct/ib/dc/base/dc_iface.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        uint8_t       *dcis_stack;
+        uint8_t       stack_top;
+        uint8_t       ndci;
+        ucs_arbiter_t dci_arbiter;
+    } tx;
+    struct { 
+        struct ibv_exp_dct *dct;
+    } rx;
+    struct {
+        int tx_policy;
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.h]: should be enum - what are possible values? if we don't use yet - lets move to next PR

File: src/uct/ib/dc/base/dc_iface.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                                              unsigned *num_resources_p);
+
+static inline uint8_t uct_dc_iface_dci_find(uct_dc_iface_t *iface, uint32_t qp_num)
+{
+    /* linear search is most probably the best way to go
+     * because the number of dcis is usually small 
+     */
+    int i;
+
+    for (i = 0; i < iface->tx.ndci; i++) {
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.h]: i would say TODO convert it to perfect hash

Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.h]: maybe even binary search (keep the dcis sorted by qpn)

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 6]
```diff
@@ -0,0 +1,771 @@
...
+    self->dest_qpn = uct_ib_unpack_uint24(if_addr->qp_num);
+    ucs_debug("created ep %p on iface %p", self, iface);
+
+    return UCS_OK;
+}
+
+static UCS_CLASS_CLEANUP_FUNC(uct_dc_verbs_ep_t)
+{
+    ucs_trace_func("");
+    if (self->ah) {
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: how can self->ah be NULL? looks like the constructor always creates it

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 7]
```diff
@@ -0,0 +1,771 @@
...
+                                      UCT_IFACE_FLAG_ATOMIC_SWAP32|
+                                      UCT_IFACE_FLAG_ATOMIC_CSWAP32|
+                                      UCT_IFACE_FLAG_PENDING |
+                                      UCT_IFACE_FLAG_AM_CB_SYNC|UCT_IFACE_FLAG_CONNECT_TO_IFACE;
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_dc_verbs_ep_create_connected(uct_iface_h iface_h, const uct_device_addr_t *dev_addr,
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: we should make this function the class constructor, and not call NEW directly.

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 8]
```diff
@@ -0,0 +1,771 @@
...
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE void
+uct_dc_verbs_iface_post_send(uct_dc_verbs_iface_t* iface, uct_dc_verbs_ep_t *ep,
+                             struct ibv_exp_send_wr *wr, uint64_t send_flags)
+{
+    struct ibv_exp_send_wr *bad_wr;
+    int ret;
+    uint8_t dci = ep->super.dci;
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: maybe better to get a pointer directly to the txqp?

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 9]
```diff
@@ -0,0 +1,771 @@
...
+    }
+    return status;
+}
+
+ucs_status_t uct_dc_verbs_ep_flush(uct_ep_h tl_ep, unsigned flags, uct_completion_t *comp)
+{
+    uct_dc_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_iface_t);
+    ucs_status_t status;
+
+    status = uct_dc_verbs_flush_dcis(iface);
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: TODO need to flush only relevant DCIs

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 10]
```diff
@@ -0,0 +1,771 @@
...
+        if (ucs_unlikely(wc[i].status != IBV_WC_SUCCESS)) {
+            ucs_fatal("iface=%p: send completion %d with error: %s wqe: %p wr_id: %llu", 
+                      iface, i, ibv_wc_status_str(wc[i].status), 
+                      &wc[i], (unsigned long long)wc[i].wr_id);
+        }
+        count = uct_rc_verbs_txcq_get_comp_count(&wc[i]);
+        ucs_assert(count == 1);
+        dci = uct_dc_iface_dci_find(&iface->super, wc[i].qp_num);
+        uct_rc_verbs_txqp_completed(&iface->super.tx.dcis[dci].txqp, &iface->dcis_txcnt[dci], count);
+        ucs_trace_poll("dc tx completion on dc %d cound %d", dci, count);
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: typo: count

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 11]
```diff
@@ -0,0 +1,771 @@
...
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    for (i = 0; i < self->super.tx.ndci; i++) {
+        uct_rc_verbs_txcnt_init(&self->dcis_txcnt[i]);
+        uct_rc_txqp_available_set(&self->super.tx.dcis[i].txqp, self->super.super.config.tx_qp_len);
+    }
+    
+    uct_worker_progress_register(worker, uct_dc_verbs_iface_progress, self);
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: need to register progress only when have an either local/remote connection, because doing progress all the time will hurt performance

File: test/gtest/uct/uct_test.h [hunk 12]
```diff
@@ -186,6 +186,7 @@ std::ostream& operator<<(std::ostream& os, const resource* resource);
 #define UCT_TEST_IB_TLS \
     rc_mlx5, \
     rc, \
+    dc, \
```
Comment (yosefe) [inline test/gtest/uct/uct_test.h]: need to add to ucp tests as well

File: src/uct/ib/dc/base/dc_ep.c [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+    ucs_arbiter_group_push_elem(&ep->arb_group, (ucs_arbiter_elem_t*)r->priv);
+    uct_dc_iface_dci_sched_tx(iface, ep);
+
+    return UCS_OK;
+}
+
+/**
+ * dispatch requests waiting for dci allocation
+ * Some dcis become available. It means that
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_ep.c]: ... ?

File: src/uct/ib/dc/base/dc_ep.h [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+ * - dcs:
+ *    - ep uses already assigned dci or
+ *    - free dci is assigned in LIFO (stack) order or
+ *    - ep has not resources to transmit
+ *    - on FULL completion (once there are no outstanding ops)
+ *      dci is pushed to the stack of free dcis
+ *    TDOO: it is possible that ep will never release its dci:
+ *      ep send, gets some completion, sends more, repeat
+ */
+#define UCT_DC_EP_NO_DCI 0xff
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_ep.h]: maybe UINT8_MAX?

File: src/uct/ib/dc/base/dc_iface.c [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+                                 IBV_EXP_ACCESS_REMOTE_READ | 
+                                 IBV_EXP_ACCESS_REMOTE_ATOMIC;
+    init_attr.min_rnr_timer    = iface->super.config.min_rnr_timer;
+    init_attr.hop_limit        = 1;
+    init_attr.inline_size      = iface->super.config.rx_inline;
+
+    iface->rx.dct = ibv_exp_create_dct(uct_ib_iface_device(&iface->super.super)->ibv_context, 
+                                       &init_attr);
+    if (iface->rx.dct == NULL) { 
+        ucs_error("Failed to created DC target");
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: pls print %m

File: src/uct/ib/dc/base/dc_iface.h (hunk 4 above)
Comment (alex-mikheev) [inline src/uct/ib/dc/base/dc_iface.h]: Only one policy DCS[+QUOTA] is implemented

Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.h]: so let's remove this for now?

File: src/uct/ib/dc/verbs/dc_verbs.c (hunk 11 above)
Comment (alex-mikheev) [inline src/uct/ib/dc/verbs/dc_verbs.c]: Putting on TODO list. It will be done in separate PR

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1403/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1404/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2027/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2028/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2030/ for details.

**Comment by alex-mikheev** [comment]
:bot:retest

**Comment by alex-mikheev** [comment]
bot:retest

**Comment by mellanox-github** [comment]
bot:retest

**Comment by Di0gen** [comment]
bot:retest

**Comment by Di0gen** [comment]
bot:retest

File: src/uct/ib/dc/base/dc_ep.h [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+    uct_dc_ep_t *ep = iface->tx.dcis[dci].ep;
+
+    ucs_assert(iface->tx.stack_top > 0);
+
+    if (uct_rc_txqp_available(&iface->tx.dcis[dci].txqp) < (int16_t)iface->super.config.tx_qp_len) {
+        if (ucs_unlikely(ep == NULL)) {
+            /* ep was destroyed while holding dci */
+            return;
+        }
+#if UCT_DC_EP_DCS_QUOTA
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_ep.h]: pls make it runtime config

**Comment by Di0gen** [comment]
bot:retest

**Comment by mellanox-github** [comment]
bot:retest

**Comment by mellanox-github** [comment]
bot:retest

**Comment by mellanox-github** [comment]
bot:retest

**Comment by mellanox-github** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2031/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2032/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2033/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2034/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2035/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2041/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1407/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2042/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2047/ for details.

File: src/uct/ib/dc/base/dc_iface.c [hunk 17]
```diff
@@ -192,9 +201,12 @@ UCS_CLASS_DEFINE(uct_dc_iface_t, uct_rc_iface_t);
 ucs_config_field_t uct_dc_iface_config_table[] = {
     {"DC_", "", NULL,
         ucs_offsetof(uct_dc_iface_config_t, super), UCS_CONFIG_TYPE_TABLE(uct_rc_verbs_iface_config_table)},
-    {"NUM_DCIS", "8", 
+    {"NUM_DCI", "8", 
      "Number of dcis",
      ucs_offsetof(uct_dc_iface_config_t, ndci), UCS_CONFIG_TYPE_UINT},
+    {"TX_POLICY", "dcs_quota",
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: pls extend help message

File: src/uct/ib/dc/base/dc_iface.c [hunk 18]
```diff
@@ -192,9 +201,12 @@ UCS_CLASS_DEFINE(uct_dc_iface_t, uct_rc_iface_t);
 ucs_config_field_t uct_dc_iface_config_table[] = {
     {"DC_", "", NULL,
         ucs_offsetof(uct_dc_iface_config_t, super), UCS_CONFIG_TYPE_TABLE(uct_rc_verbs_iface_config_table)},
-    {"NUM_DCIS", "8", 
+    {"NUM_DCI", "8",
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: pls extend help message

**Comment by yosefe** [comment]
pls also squash the commits

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2053/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1412/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2054/ for details.

**Comment by yosefe** [comment]
:+1:
